hack/generator/pkg/astmodel/kubernetes_resource_interface.go

```diff
@@ -18,44 +20,95 @@ const (
 	OwnerProperty     = "Owner"
 )
 
-// NewArmTransformerImpl creates a new interface with the specified ARM conversion functions
-func NewKubernetesResourceInterfaceImpl(
+// WithKubernetesResourceInterfaceImpl creates a new interface with the specified ARM conversion functions
+func (r *ResourceType) WithKubernetesResourceInterfaceImpl(
 	idFactory IdentifierFactory,
-	spec *ObjectType) (*InterfaceImplementation, error) {
+	types Types) (*ResourceType, error) {
+
+	resolvedSpec, err := types.FullyResolve(r.SpecType())
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to resolve resource spec type")
+	}
+
+	spec, err := TypeAsObjectType(resolvedSpec)
+	if err != nil {
+		return nil, errors.Wrapf(err, "resource spec %q did not contain an object", r.SpecType().String())
+	}
 
 	// Check the spec first to ensure it looks how we expect
 	ownerProperty := idFactory.CreatePropertyName(OwnerProperty, Exported)
 	_, ok := spec.Property(ownerProperty)
 	if !ok {
-		return nil, errors.Errorf("Resource spec doesn't have %q property", ownerProperty)
+		return nil, errors.Errorf("resource spec doesn't have %q property", ownerProperty)
 	}
 
-	azureNameProperty := idFactory.CreatePropertyName(AzureNameProperty, Exported)
-	_, ok = spec.Property(azureNameProperty)
+	nameProp, ok := spec.Property(AzureNameProperty)
 	if !ok {
-		return nil, errors.Errorf("Resource spec doesn't have %q property", azureNameProperty)
+		return nil, errors.Errorf("resource spec doesn't have %q property", AzureNameProperty)
 	}
 
-	ownerFunc := &objectFunction{
-		name:      OwnerProperty,
-		o:         spec,
-		idFactory: idFactory,
-		asFunc:    ownerFunction,
+	namePropType := nameProp.PropertyType()
+	// resolve property type if it is a typename
+	namePropType, err = types.FullyResolve(namePropType)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to resolve type of resource Name property: %s", namePropType.String())
+	}
+
+	var nameFunc asFuncType
+
+	// handle different types of Name property
+	switch namePropType := namePropType.(type) {
+	case *EnumType:
+		if !namePropType.BaseType().Equals(StringType) {
+			return nil, errors.Errorf("unable to handle non-string enum base type in Name property")
+		}
+
+		options := namePropType.Options()
+		if len(options) == 1 {
+			// if there is only one possible value,
+			// we make an AzureName function that returns it, and do not
+			// provide an AzureName property on the spec
+			//r = r.WithSpec(spec.WithoutProperty(AzureNameProperty))
```

Comment:
TODO: any ideas on how best to do this? this results in bad output as really we need to update the TypeDefinition for this Spec… and I can only come up with ugly solutions

Reply:
going with ugly solution